test/distributed/fsdp/test_composable_fsdp.py

```diff
@@ -0,0 +1,215 @@
+# Owner(s): ["oncall: distributed"]
+
+import copy
+import functools
+import sys
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp._common_utils import _is_fsdp_flattened
+from torch.distributed.fsdp._fsdp import fully_sharded_data_parallel
+from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    run_tests,
+    TEST_WITH_DEV_DBG_ASAN,
+)
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class SubModel(nn.Module):
+    def __init__(self, device) -> None:
+        super().__init__()
+        torch.manual_seed(0)
+        self.lin1 = nn.Linear(5, 5, bias=False, device=device)
+        self.lin2 = nn.Linear(5, 5, bias=False, device=device)
+        self.relu = nn.ReLU()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        z = self.relu(self.lin1(x))
+        z = self.relu(self.lin2(z))
+        return z
+
+
+class Model(nn.Module):
+    def __init__(self, device) -> None:
+        super().__init__()
+        torch.manual_seed(0)
+        self.sub1 = SubModel(device=device)
+        self.sub2 = SubModel(device=device)
+        self.lin = nn.Linear(5, 5, device=device)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        z = self.sub1(x)
+        z = self.sub2(z)
+        z = self.lin(z)
+        return z
+
+    @staticmethod
+    def auto_wrap_policy():
+        return functools.partial(
+            transformer_auto_wrap_policy, transformer_layer_cls={SubModel}
+        )
+
+
+class TestFSDPInitialization(FSDPTest):
+    """Tests composable FSDP initialization."""
+
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_auto_wrap_policy(self):
+        """Tests passing an ``auto_wrap_policy``."""
+
+        local_model = Model(device=torch.device("cuda"))
+
+        fsdp_wrapped_model = FSDP(
+            copy.deepcopy(local_model),
+            auto_wrap_policy=Model.auto_wrap_policy(),
+            use_orig_params=True,
+        )
+        composable_module = copy.deepcopy(local_model)
+        fsdp_state = fully_sharded_data_parallel(
```

Comment:
no need to change this PR either. Let's move to contract's state in follow up PRs.